File: lib/ios/Classes/LdkLogger.swift
//
//  LdkLogger.swift
//  react-native-ldk
//
//  Created by Jason van den Berg on 2022/05/10.
//

import Foundation
import LightningDevKit

fileprivate func levelString(_ level: Level) -> String {
    switch level {
    case .Gossip:
        return "GOSSIP"
    case .Trace:
        return "TRACE"
    case .Debug:
        return "DEBUG"
    case .Info:
        return "INFO"
    case .Warn:
        return "WARN"
    case .Error:
        return "ERROR"
    default:
        return "LEVEL \(level)"
    }
}

class LdkLogger: LightningDevKit.Bindings.Logger {
    var activeLevels: [String: Bool] = [:]
    
    override func log(record: Record) {
        let level = levelString(record.getLevel())
        
        //Only when the JS code has set the log level to active
        if activeLevels[level] == true {
            let line = "\(level) (LDK): \(record.getArgs()) (\(record.getModulePath()) \(record.getLine()))"
            LdkEventEmitter.shared.send(withEvent: .ldk_log, body: line)
            Logfile.log.write(line)
        }
    }
    
    func setLevel(level: String, active: Bool) {
        self.activeLevels[level] = active
        LdkEventEmitter.shared.send(withEvent: .native_log, body: "Log level \(level) set to \(active)")
    }
}

class Logfile: TextOutputStream {
    var logfile: URL?
    
    func setFilePath(_ file: URL) {
        logfile = file
    }
    
    func write(_ str: String) {
        guard let logfile = logfile else {
            return
        }
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        let line = "\(dateFormatter.string(from: Date())) \(str)\n"
        
        if let handle = try? FileHandle(forWritingTo: logfile) {
            handle.seekToEndOfFile()
            handle.write(line.data(using: .utf8)!)
            handle.closeFile()
        } else {
            try? line.data(using: .utf8)?.write(to: logfile)
        }
    }
    static var log = Logfile()
    private init() {}
}

File: lib/android/src/main/java/com/reactnativeldk/classes/LdkLogger.kt
package com.reactnativeldk.classes

import android.annotation.SuppressLint
import com.reactnativeldk.EventTypes
import com.reactnativeldk.LdkEventEmitter
import org.ldk.structs.Logger
import org.ldk.structs.Record
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date

private fun levelString(level: Int): String {
    when (level) {
        0 -> return "GOSSIP"
        1 -> return "TRACE"
        2 -> return "DEBUG"
        3 -> return "INFO"
        4 -> return "WARN"
        5 -> return "ERROR"
    }

    return "LEVEL $level"
}

class LdkLogger {
    var activeLevels = HashMap<String, Boolean>()
    var logger = Logger.new_impl{ record: Record ->
        val level = levelString(record._level.ordinal)

        if (activeLevels[level] == true) {
            val line = "$level (LDK): ${record._args} (${record._module_path} ${record._line})"
            LdkEventEmitter.send(EventTypes.ldk_log, line)
            LogFile.write(line)
        }
    }

    fun setLevel(level: String, active: Boolean) {
        activeLevels[level] = active;
        LdkEventEmitter.send(EventTypes.native_log, "Log level ${level} set to ${active}")
    }
}

object LogFile {
    private var logFile: File? = null

    fun setFilePath(logFile: File) {
        if (!logFile.isFile) {
            logFile.createNewFile()
        }

        this.logFile = logFile
    }

    @SuppressLint("SimpleDateFormat")
    fun write(str: String) {
        if (logFile == null) {
            return
        }

        val dateFormatter = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        val line = "${dateFormatter.format(Date())} $str\n"

        logFile!!.appendText(line)
    }
}
